scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+
+	if (!(lockw = bpf_cgrp_storage_get(&cgrp_locks, cgrp, 0,
+					   BPF_LOCAL_STORAGE_GET_F_CREATE))) {
+		scx_bpf_error("cgrp_lock_wrapper lookup failed for cgid %llu",
+			      cgrp->kn->id);
+		return NULL;
+	}
+
+	return lockw;
+}
+
+static inline struct cgroup *lookup_cgrp_ancestor(struct cgroup *cgrp,
+						  u32 ancestor)
+{
+	struct cgroup *cg;
+
+	if (!(cg = bpf_cgroup_ancestor(cgrp, ancestor))) {
+		scx_bpf_error("Failed to get ancestor level %d for cgid %llu",
+			      ancestor, cgrp->kn->id);
+		return NULL;
+	}
+
+	return cg;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_CGRP_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct cgrp_ctx);
+} cgrp_ctx SEC(".maps");
+
+static inline struct cgrp_ctx *lookup_cgrp_ctx(struct cgroup *cgrp)
+{
+	struct cgrp_ctx *cgc;
+
+	if (!(cgc = bpf_cgrp_storage_get(&cgrp_ctx, cgrp, 0,
+					 BPF_LOCAL_STORAGE_GET_F_CREATE))) {
+		scx_bpf_error("cgrp_ctx lookup failed for cgid %llu",
+			      cgrp->kn->id);
+		return NULL;
+	}
+
+	return cgc;
+}
+
+/* Map from cgrp -> cell populated by userspace for reconfiguration */
+struct {
+	__uint(type, BPF_MAP_TYPE_CGRP_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, u32); /* cell */
+} cgrp_cell_assignment SEC(".maps");
+
+static inline struct cgroup *task_cgroup(struct task_struct *p)
+{
+	struct cgroup *cgrp = scx_bpf_task_cgroup(p);
+	if (!cgrp) {
+		scx_bpf_error("Failed to get cgroup for task %d", p->pid);
+	}
+	return cgrp;
+}
+
+struct task_ctx {
+	struct bpf_cpumask __kptr *cell_cpumask;
+	u64 started_running_at;
+	u32 cell;
+	u32 global_seq;
+	bool all_cpus_allowed;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_TASK_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct task_ctx);
+} task_ctxs SEC(".maps");
+
+static inline struct task_ctx *lookup_task_ctx(struct task_struct *p)
+{
+	struct task_ctx *tctx;
+
+	if ((tctx = bpf_task_storage_get(&task_ctxs, p, 0, 0))) {
+		return tctx;
+	}
+
+	scx_bpf_error("task_ctx lookup failed");
+	return NULL;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctxs SEC(".maps");
+
+static inline struct cpu_ctx *lookup_cpu_ctx(int cpu)
+{
+	struct cpu_ctx *cctx;
+	u32 zero = 0;
+
+	if (cpu < 0)
+		cctx = bpf_map_lookup_elem(&cpu_ctxs, &zero);
+	else
+		cctx = bpf_map_lookup_percpu_elem(&cpu_ctxs, &zero, cpu);
+
+	if (!cctx) {
+		scx_bpf_error("no cpu_ctx for cpu %d", cpu);
+		return NULL;
+	}
+
+	return cctx;
+}
+
+struct cell {
+	u64 vtime_now;
+	u32 dsq;
+	// The following field is populated from userspace to indicate
+	// which cpus the cell should belong to.
+	unsigned char cpus[MAX_CPUS_U8];
+};
+
+struct cell cells[MAX_CELLS];
+
+static inline struct cell *lookup_cell(int idx)
+{
+	if (idx < 0 || idx >= MAX_CELLS) {
+		scx_bpf_error("invalid cell %d", idx);
+		return NULL;
+	}
+	return &cells[idx];
+}
```

Comment:
It might be less error prone / future proof to just use `MEMBER_VPTR` here:

```
struct cell *cell;

cell = MEMBER_VPTR(cells, [idx]);
if (!cell)
        scx_bpf_error("Invalid cell %d", idx);
return cell;
```